server/grpc_service.go

```diff
@@ -406,23 +406,27 @@ func (s *GrpcServer) forwardTSO(stream pdpb.PD_TsoServer) error {
 	var (
 		server            = &tsoServer{stream: stream}
 		forwardStream     tsopb.TSO_TsoClient
-		cancel            context.CancelFunc
+		forwardCtx        context.Context
+		cancelForward     context.CancelFunc
 		lastForwardedHost string
 	)
 	defer func() {
 		s.concurrentTSOProxyStreamings.Add(-1)
-		// cancel the forward stream
-		if cancel != nil {
-			cancel()
+		if cancelForward != nil {
+			cancelForward()
 		}
 	}()
+
 	maxConcurrentTSOProxyStreamings := int32(s.GetMaxConcurrentTSOProxyStreamings())
 	if maxConcurrentTSOProxyStreamings >= 0 {
 		if newCount := s.concurrentTSOProxyStreamings.Add(1); newCount > maxConcurrentTSOProxyStreamings {
 			return errors.WithStack(ErrMaxCountTSOProxyRoutinesExceeded)
 		}
 	}
 
+	tsDeadlineCh := make(chan *tsoutil.TSDeadline, 1)
```

Comment:
how about using `context.withTimeout(forwardCtx,...)`

Reply:
We can't use context.withTimeout() per grpc.ServerStream, because here we want to add timeout mechanism to watch every request forwarded by the forward stream instead of the entire grpc.ServerStream's deadline. The existing logic use context.withTimeout() then asynchronously call forwardTSORequest which is expensive and the latency varies in a wider range due to synchronization across multiple goroutines. This change aims to fix this problem as described in the PR description "What is changed and how does it work?" which is also pasted below:

_In order to handle the timeout case for forwardStream send/recv, the existing logic is to create 
context.withTimeout(forwardCtx,...) for every request, then start a new goroutine "forwardTSORequest", 
which is very expensive as shown by the profiling in #6659. 

This change create a watchDeadline routine per forward stream and reuse it for all the forward requests
in which forwardTSORequest is called synchronously. Compared to the existing logic, the new change
is much cheaper and the latency is much stable._